### ISSUE TITLE
Select API/documentation is somewhat misleading

### ISSUE THREAD

**Issue by cBournhonesque** [coreylowman/dfdx#296]
I was looking at the docs for the 'Select' API and I was a bit confused.
https://github.com/coreylowman/dfdx/blob/main/src/tensor_ops/select.rs#L40

```
/// // select a single element from the 1st axis - number of elements is equal
/// // to the size of the 0th axis, and the usize values can be 0..5
/// let _: Tensor1D<3> = Tensor2D::<3, 5>::zeros().select(&[0, 2, 4]);
```

I think the language 'select a single element from the 1st axis' followed by `number of elements is equal to the size of the 0th axis' is a bit confusing. I remember thinking: so are we selecting a single element or not? Is it a single column?

Looking at the API, it looks like there's no way to way to select a 'column' by providing a single index, like with Tensorflow or Pytorch?
i.e. if I want to select the 2nd column (0-indexed), I would have to write
`let _: Tensor1D<3> = Tensor2D::<3, 5>::zeros().select(&[2, 2, 2]);`
instead of something like
`let _: Tensor1D<3> = SelectTo<_, Axis<1>>::select(Tensor2D::<3,5>::zeros(), 2);`
or
`let _: Tensor1D<3> = Tensor2D::<3, 5>::zeros().select(2, axis=Axis<1>);`

With higher dimensional tensors, it gets even more complicated:
`impl_select!(Axis<2>, SelectAx2, Tensor4D<M, N, O, P, H>, [[[usize; Z]; N]; M], Tensor4D<M, N, Z, P, H>, {M, N, O, P, Z});`
It looks like I would need to pass `[[[usize; Z]; N]; M]` as input to select elemnts from Axis2.

I'm probably missing something or misunderstanding; but if we're passing the Axes as generics in the `SelectTo<T, Axes>` trait, why do the `indices` input need to change depending on the axes we're selecting?

**Comment by coreylowman** [comment]
Yeah I see how it's confusing, thanks for raising this. Revisiting this I think we can just accept either only `usize` or `[usize; Z]` and then just "broadcast" the indices to be the correct size at device level. This may actually be pretty simple:

I think we can just change `[I; M]` to `I` on this line: https://github.com/coreylowman/dfdx/blob/main/src/devices/select.rs#L101
```diff
-impl<T, I, const M: usize, SubMode> DeviceSelect<[T; M], [I; M], Recurse<SubMode>> for Cpu
+impl<T, I, const M: usize, SubMode> DeviceSelect<[T; M], I, Recurse<SubMode>> for Cpu
```

And then update the types for the indices in tensor_ops macros.

# 
For select there's kind of two different modes:
1. select a single item
2. select multiple items

e.g. if I have `Tensor2D<3, 5>`, I could select a single item from the 0th axis to get a `Tensor1D<5>`, or a single item from the 1st axis to get a `Tensor1D<5>`.

So selecting a single item is dimension reducing - it removes one of the dimensions.

Multiple items however keeps the dimensions the same. If I select 2 items from 0th axis, I get a `Tensor2D<2, 5>`, and from the 1st axis a `Tensor2D<3, 2>`.

Hopefully after the above simplifications to the indices this becomes more clear? Does the above make sense? Does it come through in the existing documentation?

**Comment by coreylowman** [comment]
Ahh now I remember why I did the `[[[usize; Z]; N]; M]` stuff 😅 - it supports a bit of a more general case than just passing in `[usize; Z]`.

If i have a `Tensor2D<2, 3>` and I wanted to select the first item from the 2nd axis. Something like this could work:
```rust
let a: Tensor2D<2, 3> = tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]);
// NOTE: this doesn't work currently
let b: Tensor1D<2> = a.select(&0);
assert_eq!(b.data(), &[1.0, 4.0]);
```

but what if i wanted to select different things based on the 1st axis? Like I wanted the `1.0` from the first sub array, and `5.0` from the second sub array. That's where the current implementation lands:

```rust
let a: Tensor2D<2, 3> = tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]);
let b: Tensor1D<2> = a.select(&[0, 1]);
assert_eq!(b.data(), &[1.0, 5.0]);
```

Does that make sense? I think we could support both tbh, but unclear what the main use cases for people are

**Comment by cBournhonesque** [comment]
Thanks for the explanations!
Yep I get that the `[[[usize; Z]; N]; M]` was to be able to select different things based on the 1st axis. I can still select an entire column `let _: Tensor1D<3> = Tensor2D::<3, 5>::zeros().select(&[2, 2, 2]);` but it's a bit unwiedly..

I would say that most of the time, people want to be able to select/slice through a certain axis (for example select a slice from the 2nd axis). So if it's not hard to add, supporting both would be worth it!

**Comment by cBournhonesque** [comment]
I've tried adding the new api for selecting an entire column which is not the first axis: https://github.com/coreylowman/dfdx/commit/7359bec566aba15d85652e5478e414ee5c0d58a4#diff-551f50e31b857028780e4943c1af4254e153a5f73ba4c267e225aa01eb618eafR109

However I have 2 issues:
- i can't figure out how to initialize the indices array within the declarative macro (`[[indices; N]; M]`)
- for some reason, once I add this extra impl, rust complains and says that extra type annotations are needed. This seems surprising to me because it looks like the types should be able to get inferred

**Comment by coreylowman** [comment]
I'm hoping some changes I introduced in #304 help with this confusion. Notably select has been split into two different functions:
1. select - select a single value from a single axis
2. gather - select multiple values from a single axis

You'll still need to provide a multi-dimensional index tensor, but i'm hoping the new documentation makes it more clear on what those need to look like.

Additionally, since the indices are now tensors with usize dtype, you can use broadcast on them if you really just have a single index, e.g
```rust
let t: Tensor<Rank3<3, 4, 5>, f32> = dev.zeros();

// select index 0 from axis 1 - removing axis 1 from shape
let _: Tensor<Rank2<3, 5>, f32> = t.select(dev.tensor([0, 0, 0]));

// OR just use broadcast
let _: Tensor<Rank2<3, 5>, f32> = t.select(dev.tensor(0).broadcast());
```

**Comment by coreylowman** [comment]
Closing this for now. @cBournhonesque check out the new index example and the docstrings for select/gather under tensor_ops. If they are still confusing let me know and we can re-open.
